CMake: on Darwin be explicit about inclusion in the dyld shared cache

edymtt opened the pull request:
This entails passing a linker flags to Apple linkers when the standard library is not meant for inclusion in such cache.

For this to have effect on every library, propagate link flags when building _Concurrency and Observation.

This is needed for Apple internal configurations.

Addresses rdar://120653968

edymtt commented on stdlib/public/CMakeLists.txt:
```diff
@@ -19,6 +19,20 @@ if(SWIFT_STDLIB_SIL_DEBUGGING)
+       FALSE)
+
+if(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
+    set(linker_detected "$<BOOL:${LLVM_LINKER_DETECTED}>")
```
For `LLVM_LINKER_DETECTED` to catch properly all the supported Apple linkers, we would need to amend the linker check for macOS -- I have a [LLVM PR for this](https://github.com/llvm/llvm-project/pull/77806)

edymtt commented:
@swift-ci please test

edymtt commented:
@swift-ci please build toolchain

etcwilde reviewed (commented):
Do we need to use the generator expressions here? I might be missing something so correct me if I'm wrong, but it looks like these are configuration-time booleans? If it is, can we stick with `if` statements? They're easier to debug since they show up in the output of `--trace-expand` instead of having to track it down in the `build.ninja` file.

edymtt commented:
@swift-ci please test

edymtt commented:
@swift-ci please build toolchain

edymtt commented:
I used generator expressions based on the feedback on a [past PR that dealt with linker flags](https://github.com/apple/swift/pull/41058), assuming it was the preferred approach going forward -- likely I missed the context there and thought I could apply that verbatim here.

Reworked this in [fc60c6c](https://github.com/apple/swift/pull/70856/commits/fc60c6c7b53fbc790dd7d7e7498b0943b6850809)

To answer your second question: yes, I'm using configuration-time booleans.

etcwilde commented on stdlib/public/CMakeLists.txt:
```diff
@@ -19,6 +19,19 @@ if(SWIFT_STDLIB_SIL_DEBUGGING)
+       FALSE)
+
+if(CMAKE_SYSTEM_NAME STREQUAL "Darwin" AND
+   LLVM_LINKER_DETECTED AND
```
Based on https://github.com/llvm/llvm-project/pull/77806/files#diff-f60dae8e82f67fa5e534e9ea566430d5b79f36119020865dc37b95d9a3635c5eR216, it seems like this should be based on `LLVM_LINKER_IS_APPLE` instead.

edymtt replied:
Good observation -- in hindsight I wanted to be too clever, the flag is only properly understood by Apple linkers (LLD in fact ignores it).

Feedback taken in abac6822b2bf567e39f29c40b9e326024dd6c738 (and reworded commit description appropriately)

edymtt commented on stdlib/public/CMakeLists.txt:
```diff
@@ -19,6 +19,19 @@ if(SWIFT_STDLIB_SIL_DEBUGGING)
+       FALSE)
+
+if(CMAKE_SYSTEM_NAME STREQUAL "Darwin" AND
+   LLVM_LINKER_IS_APPLE AND
```
`LLVM_LINKER_IS_APPLE` will be introduced with https://github.com/llvm/llvm-project/pull/77806.

Keeping this as a draft PR for now until I understand the outcome of the review of the LLVM one.

edymtt commented:
@swift-ci please test

edymtt commented:
@swift-ci please build toolchain

edymtt commented:
@swift-ci please test

edymtt commented:
@swift-ci please build toolchain

Azoy reviewed (approved):
Looks good to me, does this also need to add `LINK_FLAGS` to the `Synchronization` module?

compnerd commented on stdlib/public/CMakeLists.txt:
```diff
@@ -19,6 +19,19 @@ if(SWIFT_STDLIB_SIL_DEBUGGING)
   list(APPEND SWIFT_RUNTIME_SWIFT_COMPILE_FLAGS "-Xfrontend" "-sil-based-debuginfo")
 endif()
 
+option(SWIFT_RUNTIME_PART_OF_DARWIN_SHARED_CACHE
```
Personally, I think that `SWIFT_RUNTIME_DYLD_SHARED_CACHE_COMPATIBLE` is a better option name.

edymtt replied:
Yes, I agree -- it makes the code more readable, since there is a visual match for "dyld shared cache" in the condition and the flag applied.

Applied this suggestion in [d84140d](https://github.com/apple/swift/commit/d84140d1d331b7c665bb8ea72c79be297191444d)

edymtt committed:
CMake: on Darwin be explicit about inclusion in the dyld shared cache

This entails passing a linker flags to Apple linkers when the standard
library is not meant for inclusion in such cache.

For this to have effect on every library, propagate link flags when
building _Concurrency and Observation.

This is needed for Apple internal configurations.

Addresses rdar://120653968

edymtt commented:
Good point, I totally forgot to check Synchronization when marking this PR as ready to review -- tackling that in https://github.com/apple/swift/pull/70856/commits/d84140d1d331b7c665bb8ea72c79be297191444d (and adding a missing newline while I'm at it)

edymtt commented:
@swift-ci please test

edymtt commented:
@swift-ci please build toolchain

edymtt commented:
@swift-ci please build toolchain Windows
